Reference/Common-Pitfalls/index-v9.md

```diff
@@ -0,0 +1,426 @@
+
+        // Just return the Ids, they can be resolved to IPublishedContent instances in the view or elsewhere,
+        // doesn't need to be in the model - this would also be bad if the model was cached since all of the
+        // related entities would end up in the cache too.
+        private List<int> _related;
+
+        public IEnumerable<int> RelatedRecipes => _related ??= this.Value<IEnumerable<int>>("related").ToList();
+    }
+```
+
+This is slightly better:
+
+* This will iterate over all Recipes, create and allocate 5000 instances of `IPublishedContent`
+* This will create and allocate 5000 instances of `RecipeModel`
+
+This means that there is now a minimum of __15,000__ new objects created and allocated in memory. The number of traversals/visits to each
+of these objects is now __5000__.
+
+This is still not great though. There really isn't much reason to create a `RecipeModel` to use it as a filter,
+this is allocating a lot of objects to memory for no real reason. This could be written like:
+
+```csharp
+@{
+	var recipeNode = Umbraco.Content(1234);
+}
+
+<ul>
+	@foreach (var recipe in recipeNode.Children
+		.OrderByDescending(x => x.Value<int>("votes"))
+		.Take(10))
+	{
+		<li><a href="@recipe.Url()">@recipe.Name</a></li>
+	}
+</ul>
+```
+
+This is slightly better:
+
+This means that there is now a minimum of __10,000__ new objects created and allocated in memory. The number of traversals/visits to each
+of these objects is now __5000__.
+
+## Not caching epxensive lookups
+
+Based on the above 2 points, you can see that iterating content with the traversal APIs will cause new 
+instances of `IPublishedContent` to be created. When memory is used, Garbage Collection needs to occur and this
+turnover can cause performance problems. The more objects created, the more items allocated in memory, the harder the job is for the Garbage Collector == more performance problems. Even worse is when you allocate tons of items in memory and/or really large items in memory. They will remain in memory for a long time because they'll end up in something called "Generation 3" which the GC tries to ignore for as long as possible. It does so because it knows it's going to take a lot of resources to cleanup!
+
+
+So what can we do to mitigate this? Unfortunately there is no silver bullet that will solve all your performance issues, it will always depend on your specific scenario. One great tip though is to cache the IDs of the content you need in your critical code, and then retrieve it from the cache by ID. For instance if you need to render the same four pieces of content for your nav, then cache, or simply hardcore, the IDs of the content items and retrieve them with the ID using `Umbraco.Content`, this will always be much, much faster than trying to traverse your content tree and finding the content programatically, since it will do a direct lookup in the cache, meaning that only the `IPublishedContent` you actually need will be created.
```

Comment:
```suggestion
So what can we do to mitigate this? Unfortunately, there is no silver bullet that will solve all your performance issues, it will always depend on your specific scenario. One great tip though is to cache the IDs of the content you need in your critical code and then retrieve it from the cache by ID. For instance, if you need to render the same four pieces of content for your nav, then cache, or simply hardcore, the IDs of the content items and retrieve them with the ID using `Umbraco.Content`, this will always be much, much faster than trying to traverse your content tree and finding the content programmatically, since it will do a direct lookup in the cache, meaning that only the `IPublishedContent` you actually need will be created.
```